torch/csrc/perf_tools/ThroughputBenchmark.h

```diff
@@ -0,0 +1,169 @@
+
+#include <ATen/core/ivalue.h>
+#include <torch/csrc/jit/script/module.h>
+#include <pybind11/pybind11.h>
+
+#include <vector>
+#include <memory>
+
+namespace py = pybind11;
+
+namespace torch {
+namespace throughput_benchmark {
+
+/**
+ * The struct is used to provide results of a benchmark to the caller
+ * In the future all additional statics should be added here. This struct
+ * is exposed as an object to python with the following read-only properties:
+ *
+ * stats.qps
+ * stats.num_forwards
+ * stats.latency_avg_ms
+ *
+ */
+struct TORCH_API BenchmarkExecutionStats {
+  float latency_avg_ms{-1};
+  int64_t num_iters{-1};
+};
+
+/**
+ * Use this struct in order to configure a throughput benchmark run.
+ * This struct should include parameters related to threading, batching, number
+ * of iterations, warm-up, etc. More configs can be added as needed.
+ * General rule here is that only things that c++ must(!) to be aware of should
+ * be here. If we can keep other parts in python, we should keep them there.
+ * This is typical for things that are not perf critical and don't affect
+ * execution statistics benchmark returns.
+ */
+struct TORCH_API BenchmarkConfig {
+ public:
+  // Calling threads are those threads that are calling into a module in
+  // parallel.
+  int num_calling_threads{1};
+  // Worker threads are not supported yet. This is just an example that we plan
+  // to support some sort of multi-threaded forward calls. We may change this
+  // setting in the future to support different intra and inter op parallelizm
+  // which is not available in PyTorch yet
+  int num_worker_threads{1};
+  // Warmup iters are used to make sure we run a module a few times before
+  // actually measuring things. This way we avoid cold caches and any other
+  // similar problems
+  int num_warmup_iters{1};
+  // Number of iterations the benchmark should run with. This number is separate
+  // from the warmup iterations
+  int64_t num_iters{100};
+};
+
+namespace detail {
+
+/**
+ * A helper class to abstract out different models we test throughput of
+ */
+template <class Input, class Output, class Model>
+class BenchmarkHelper {
+public:
+  BenchmarkHelper(): initialized_{false} {}
+  explicit BenchmarkHelper(Model model): model_(model), initialized_(true) {}
+  // This method to be used in benchmark() method
+  Output runOnce(Input) const;
+  // This method is to be used when calling from Python dirrectly
+  Output runOnce(py::args, py::kwargs) const;
+  // Aggregate input in the format Model expects in order to avoid further
+  // conversions at the benchmark time
+  void addInput(py::args, py::kwargs);
+  BenchmarkExecutionStats benchmark(const BenchmarkConfig& config) const;
+
+  bool initialized() const { return initialized_; }
+
+  std::vector<Input> inputs_;
+  Model model_;
+  bool initialized_{false};
+};
+
+struct TORCH_API ModuleInput {
+  ModuleInput(ModuleInput&& other) = default;
+  ModuleInput(const ModuleInput&) = default;
```

Comment:
maybe just make it move-only? then it's "safer" to manipulate it outside of GIL

also, py::args is a py::tuple and is a py::object. Move constructor doesn't touch refcount: https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L238 - you should be able to remove a lot of GIL-iness

Reply:
hmmmm. As I mentioned above, I got pretty unlucky with GIL around runOnce(). But you are right, it should not touch it. Very strange, will check it out.

Reply:
OMG, all this time it was destructor of the result, not the move itself! So my very first version I did here got the GIL right, I just should have not returned anything here.